### PR TITLE
Snapshot bot enhancements

### PR DESCRIPTION
**Pull request by alexdewar**
This PR implements a few changes suggested by @aphilippides. It's based on #293, so we should merge that first.

The two main changes are:
1. Do training after recording images so we can record at the camera's maximum frame rate (#287)
2. Add option to record databases as video files (useful for larger datasets, #289)

I also ended up refactoring some ``ImageDatabase`` code so the image and video recording classes are now virtual, but no functional change was intended there.

### CODE REVIEW

**Comment by alexdewar** [comment]
I've put this as a draft because I've still got to test that it works on the actual robot.

File: include/navigation/image_database.h [hunk 1]
```diff
@@ -374,29 +387,24 @@ class ImageDatabase
...
         void record(const cv::Mat &image, Entry entry)
         {
-            this->addEntry(image, entry);
+            this->addEntry(image, entry, getFileName);
         }
 
-        std::string getCurrentFilenameRoot() const override
-        {
+    private:
+        const std::function<std::string()> getFileName = [this]() {
```
Comment (neworderofjamie) [inline include/navigation/image_database.h:401]: This seems a very odd use of ``std::function``. Why not make this a normal function and bind it when you assign it to a ``std::function`` with ``std::bind(&RouteRecorder::getFileName, this);``

File: projects/snapshot_bot/config.cc [hunk 2]
```diff
@@ -0,0 +1,271 @@
+#include "config.h"
```
Comment (neworderofjamie) [inline projects/snapshot_bot/config.cc:1]: Splitting that awful header into a cc is an excellent move and must do wonders for the compile time...

File: projects/snapshot_bot/config.cc [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        cv::read(node["bestSnapshotServerListenPort"], m_BestSnapshotServerListenPort, m_BestSnapshotServerListenPort);
+        cv::read(node["autoTrain"], m_AutoTrain, m_AutoTrain);
+        cv::read(node["moveSpeed"], m_MoveSpeed, m_MoveSpeed);
+
+        double trainInterval;
+        cv::read(node["trainInterval"], trainInterval, m_TrainInterval.count());
+        m_TrainInterval = (Milliseconds)trainInterval;
+
+        int frameSkip = m_SkipFrames;
+        cv::read(node["frameSkip"], frameSkip, frameSkip);
```
Comment (neworderofjamie) [inline projects/snapshot_bot/config.cc:154]: might be worth making clear exactly when these frames are skipped i.e ``m_TestSkipFrames``

File: projects/snapshot_bot/config.cc [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+// Standard C++ includes
+#include <limits>
+
+using namespace BoBRobotics;
+using Milliseconds = std::chrono::duration<double, std::milli>;
+
+namespace {
+// **YUCK** as well as std::strings, it also seems like OpenCV can't serialise size_t
+int getIntegerSize(size_t value)
```
Comment (neworderofjamie) [inline projects/snapshot_bot/config.cc:15]: you could just accept that the OpenCV serialization sucks somewhat, use ``int``/``unsigned int`` and ``static_cast`` when required..

**Review by neworderofjamie — COMMENTED** [review]
This generally all looks very sensible. Just because of the large number of changes, personally I would be _very_ reticent about using this version for the showcase but up to you!

File: include/navigation/image_database.h (hunk 1 above)
Comment (alexdewar) [reply include/navigation/image_database.h:401]: Ohh good idea. I didn't know about ``std::bind``.

File: projects/snapshot_bot/config.cc (hunk 3 above)
Comment (alexdewar) [reply projects/snapshot_bot/config.cc:154]: :+1:

**Pull request merged**